test/test_lazy.py

```diff
@@ -0,0 +1,231 @@
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+from functools import partial
+from pytools.obj_array import make_obj_array
+import pyopencl as cl
+import pyopencl.tools as cl_tools
+import pyopencl.array as cla  # noqa
+import pyopencl.clmath as clmath  # noqa
+from meshmode.array_context import (  # noqa
+    PyOpenCLArrayContext,
+    PytatoPyOpenCLArrayContext
+)
+from arraycontext.container.traversal import freeze, thaw
+
+from meshmode.array_context import (  # noqa
+    pytest_generate_tests_for_pyopencl_array_context
+    as pytest_generate_tests)
+
+import pytest  # noqa
+
+import logging
+logger = logging.getLogger(__name__)
+
+
+@pytest.fixture
+def op_test_data(ctx_factory):
+    cl_ctx = ctx_factory()
+    queue = cl.CommandQueue(cl_ctx)
+
+    eager_actx = PyOpenCLArrayContext(
+        queue,
+        allocator=cl_tools.MemoryPool(cl_tools.ImmediateAllocator(queue)))
+
+    lazy_actx = PytatoPyOpenCLArrayContext(
+        queue,
+        allocator=cl_tools.MemoryPool(cl_tools.ImmediateAllocator(queue)))
+
+    def get_discr(order):
+        from meshmode.mesh.generation import generate_regular_rect_mesh
+        mesh = generate_regular_rect_mesh(
+            a=(-0.5,)*2,
+            b=(0.5,)*2,
+            nelements_per_axis=(4,)*2,
+            boundary_tag_to_face={
+                "x": ["-x", "+x"],
+                "y": ["-y", "+y"]
+            })
+
+        from grudge.eager import EagerDGDiscretization
+        return EagerDGDiscretization(eager_actx, mesh, order=order)
+
+    return eager_actx, lazy_actx, get_discr
+
+
+def _rel_linf_error(discr, x, y):
+    from mirgecom.fluid import ConservedVars
+    if isinstance(x, ConservedVars):
+        return _rel_linf_error(discr, x.join(), y)
+    if isinstance(y, ConservedVars):
+        return _rel_linf_error(discr, x, y.join())
+    return discr.norm(x - y, np.inf) / discr.norm(y, np.inf)
```

Comment:
@majosm do you trust this comparison for CV (or other multi-component solution vectors)?  It seems to compare them as a whole, but each component has its own equation and expected result. I'm thinking about a situation similar to this:

```
x = < .01, 20, 30, 40, 50>
y = < .06, 20, 30, 40, 50>
```

In this case, the first equation just failed to get the same result between the two methods (~83% error), but the relative error comes out to only 1e-3.

Another concern about this comparison is that it is comparing the maximum difference to the maximum value over the whole domain - meaning the error at any given point could be quite large - but if it is small compared to the domain-wide max value of the fields, it is lost.

How would you feel about replacing this with something that evaluates the maximum (relative) error in each component across the whole domain?

```
perr = x - y / y   # relative error in each field at each point
max_err = [np.max(perr[i]) for i in range(len(perr))]  # max relerr in each field
```
Of course, there would be trouble if your expected field was anywhere zero (and with your init below, y will have zeros).

Are these concerns misguided?

Reply:
> @majosm do you trust this comparison for CV (or other multi-component solution vectors)? It seems to compare them as a whole, but each component has its own equation and expected result. I'm thinking about a situation similar to this:
> 
> ```
> x = < .01, 20, 30, 40, 50>
> y = < .06, 20, 30, 40, 50>
> ```
> 
> In this case, the first equation just failed to get the same result between the two methods (~83% error), but the relative error comes out to only 1e-3.

Agreed, and fixed (5c7ac75).

> Another concern about this comparison is that it is comparing the maximum difference to the maximum value over the whole domain - meaning the error at any given point could be quite large - but if it is small compared to the domain-wide max value of the fields, it is lost.
> 
> How would you feel about replacing this with something that evaluates the maximum (relative) error in each component across the whole domain?
> 
> ```
> perr = (x - y) / y   # relative error in each field at each point
> max_err = [np.max(perr[i]) for i in range(len(perr))]  # max relerr in each field
> ```
> 
> Of course, there would be trouble if your expected field was anywhere zero (and with your init below, y will have zeros).

Yeah, the zero/near-zero edge cases make this difficult. I don't think it's too unreasonable to just use a single global scaling factor per component here. For the most part we can pick and choose which cases to use in tests, so we can generally avoid those that have too large a range of scales.